crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -534,6 +543,157 @@ impl kitsune2_api::KitsuneHandler for HolochainP2pActor {
             }
         })
     }
+
+    fn preflight_gather_outgoing(&self, _peer_url: Url) -> K2Result<bytes::Bytes> {
+        Ok(self.preflight.lock().unwrap().clone())
+    }
+
+    fn preflight_validate_incoming(&self, _peer_url: Url, data: bytes::Bytes) -> K2Result<()> {
+        // decode the preflight that the remote sent us
+        let rem = crate::wire::WirePreflightMessage::decode(&data)
+            .map_err(|err| K2Error::other_src("Invalid remote preflight", err))?;
+
+        // if the compats don't match, reject the connection
+        if rem.compat != self.compat {
+            return Err(K2Error::other(format!(
+                "Invalid remote preflight, wanted {:?}, got: {:?}",
+                self.compat, rem.compat
+            )));
+        }
+
+        let mut agents = Vec::new();
+
+        // decode the agents inline, so we can reject the connection
+        // if they sent us bad agent data
+        for agent in rem.agents {
+            agents.push(AgentInfoSigned::decode(
+                &kitsune2_core::Ed25519Verifier,
+                agent.as_bytes(),
+            )?);
+        }
+
+        // if they sent us agents, spawn a task to insert them into the store
+        if !agents.is_empty() {
+            let kitsune = self.kitsune.clone();
+            tokio::task::spawn(async move {
+                for agent in agents {
+                    let space = match kitsune.space_if_exists(agent.space.clone()).await {
+                        None => continue,
+                        Some(space) => space,
+                    };
+                    space.peer_store().insert(vec![agent]).await?;
+                }
+                K2Result::Ok(())
+            });
+        }
+
+        Ok(())
+    }
+}
+
+/// A wrapper for the default Bootstrap K2 module, so we can capture
+/// the generated agent infos for publishing in the preflight.
+#[derive(Debug)]
+struct BootWrap {
+    compat: NetworkCompatParams,
+    preflight: Arc<std::sync::Mutex<bytes::Bytes>>,
+    orig: kitsune2_api::DynBootstrap,
+    cache: std::sync::Mutex<Vec<Arc<AgentInfoSigned>>>,
+}
+
+impl kitsune2_api::Bootstrap for BootWrap {
+    fn put(&self, info: Arc<AgentInfoSigned>) {
+        let Self {
+            compat,
+            preflight,
+            orig,
+            cache,
+        } = self;
+
+        let agents = {
```

Comment:
Would it be wrong to depend on the local peer store and query that instead of implementing a new memory cache here?

Reply:
I see, yes okay I can see why this needs to work this way then